File: problems/2816/listnode.ts
/**
 * Definition for singly-linked list.
 */
export class ListNode {
  val: number
  next: ListNode | null
  constructor(val?: number, next?: ListNode | null) {
    this.val = (val === undefined ? 0 : val)
    this.next = (next === undefined ? null : next)
  }
}

export function array2ListNode(values: number[]): ListNode | null {
  if (values.length === 0) return null;
  const head = new ListNode(values[0], null);
  let node: ListNode | null = head;
  for (let idx = 1, length = values.length; idx < length; idx++)
    node = node.next = new ListNode(values[idx], null);
  return head;
};

export function ListNode2array(head: ListNode | null): number[] {
  const values: number[] = [];
  for (let node: ListNode | null = head; node != null; node = node.next)
    values.push(node.val);
  return values;
};

export default {
  ListNode,
  array2ListNode,
  ListNode2array,
};

File: problems/2816/ListNode.java
import java.util.ArrayList;

/**
 * Definition for singly-linked list.
 */
public class ListNode {
  int val;
  ListNode next;

  ListNode() {
  }

  ListNode(int val) {
    this.val = val;
  }

  ListNode(int val, ListNode next) {
    this.val = val;
    this.next = next;
  }

  static ListNode fromArray(int[] values) {
    ListNode head = null;
    if (values.length > 0) {
      ListNode node = head = new ListNode(values[0]);
      for (int pos = 1; pos < values.length; ++pos)
        node = node.next = new ListNode(values[pos]);
    }
    return head;
  }

  int[] toArray() {
    ArrayList<Integer> valueIntegers = new ArrayList<>();
    for (ListNode node = this; node != null; node = node.next)
      valueIntegers.add(node.val);
    int[] values = new int[valueIntegers.size()];
    for (int pos = 0; pos < values.length; ++pos)
      values[pos] = valueIntegers.get(pos);
    return values;
  }
}
